### PR TITLE
evm_rpc

### PR DESCRIPTION
**Pull request by gtygo**

### CODE REVIEW

File: core/txlogic/tx_generate.go [hunk 1]
```diff
@@ -193,6 +193,30 @@ func MakeUnsignedTx(
 	return tx, nil
 }
 
+//MakeUnsignedContractTx make a contract tx without signature
+func MakeUnsignedContractTx(addr string, amount uint64, changeAmt uint64, gasLimit uint64,
```
Comment (Jerick26) [inline core/txlogic/tx_generate.go]: refer to MakeUnsignedSplitAddrTx format

File: core/txlogic/tx_generate.go [hunk 2]
```diff
@@ -193,6 +193,30 @@ func MakeUnsignedTx(
...
+	gasPrice uint64, byteCode []byte, utxos ...*rpcpb.Utxo,
+) (*types.Transaction, error) {
+
+	if !checkAmount(append(make([]uint64, 1), amount), changeAmt, utxos...) {
```
Comment (Jerick26) [inline core/txlogic/tx_generate.go]: bug here

File: core/txlogic/tx_generate.go (hunk 1 above)
Comment (Jerick26) [inline core/txlogic/tx_generate.go]: function name is ambiguous

File: rpc/pb/transaction.proto [hunk 3]
```diff
@@ -60,26 +60,41 @@ service TransactionCommand {
...
+        };
+    }
+
+}
+message MakeContractTxReq {
+    string addr = 1;
+    uint64 amount = 3;
+    uint64 gas_price = 4;
+    uint64 gas_limit = 5;
+    string bytecode = 6;
```
Comment (Jerick26) [inline rpc/pb/transaction.proto]: bytecode to code

File: rpc/pb/transaction.proto [hunk 4]
```diff
@@ -60,26 +60,41 @@ service TransactionCommand {
...
+        option (google.api.http) = {
+            post: "/v1/tx/makeunsignedtx/contract"
+            body: "*"
+        };
+    }
+
+}
+message MakeContractTxReq {
+    string addr = 1;
+    uint64 amount = 3;
```
Comment (Jerick26) [inline rpc/pb/transaction.proto]: 2???

File: rpc/rpcutil/transaction.go [hunk 5]
```diff
@@ -479,3 +502,17 @@ func calcChangeAmount(amounts []uint64, gasUsed uint64, utxos ...*rpcpb.Utxo) (u
 	changeAmt := uv - total
 	return changeAmt, changeAmt > uv
 }
+
+func calcChangeContractAmount(amount uint64, gasUsed uint64, utxos ...*rpcpb.Utxo) (uint64, bool) {
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: this function is unnecessary

File: rpc/pb/transaction.proto (hunk 4 above)
Comment (gtygo) [reply rpc/pb/transaction.proto]: done

File: rpc/pb/transaction.proto (hunk 3 above)
Comment (gtygo) [reply rpc/pb/transaction.proto]: done

File: rpc/rpcutil/transaction.go (hunk 5 above)
Comment (gtygo) [reply rpc/rpcutil/transaction.go]: done

File: core/txlogic/tx_generate.go (hunk 1 above)
Comment (gtygo) [reply core/txlogic/tx_generate.go]: done

File: core/txlogic/tx_generate.go (hunk 2 above)
Comment (gtygo) [reply core/txlogic/tx_generate.go]: done

File: core/txlogic/tx_generate.go [hunk 6]
```diff
@@ -193,6 +193,59 @@ func MakeUnsignedTx(
 	return tx, nil
 }
 
+//MakeUnsignedContractDeployTx make a contract tx without signature
+func MakeUnsignedContractDeployTx(
+	addr string, amount uint64, changeAmt uint64, gasLimit uint64,
+	gasPrice uint64, byteCode []byte, utxos ...*rpcpb.Utxo,
+) (*types.Transaction, error) {
+
+	amounts := append(make([]uint64, 0), amount)
```
Comment (Jerick26) [inline core/txlogic/tx_generate.go]: amounts := []uint64{amount}

File: rpc/server/rpc_transaction.go [hunk 7]
```diff
@@ -443,3 +443,82 @@ func MakeTxRawMsgsForSign(tx *types.Transaction, utxos ...*rpcpb.Utxo) ([][]byte
...
+		pbTx, err := tx.ConvToPbTx()
+		if err != nil {
+			return newMakeContractTxResp(-1, err.Error(), nil, nil, ""), err
+		}
+		rawMsgs, err := MakeTxRawMsgsForSign(tx, utxos...)
+		if err != nil {
+			return newMakeContractTxResp(-1, err.Error(), nil, nil, ""), err
+		}
+		return newMakeContractTxResp(0, "success", pbTx, rawMsgs, contractAddr.String()), nil
+	} else if req.GetContractAddr() != "" {
```
Comment (Jerick26) [inline rpc/server/rpc_transaction.go]: lint error, no need to add if clause.

File: rpc/server/rpc_transaction.go [hunk 8]
```diff
@@ -443,3 +443,82 @@ func MakeTxRawMsgsForSign(tx *types.Transaction, utxos ...*rpcpb.Utxo) ([][]byte
...
 }
+
+//MakeUnsignedContractTx make contract tx
+func (s *txServer) MakeUnsignedContractTx(
+	ctx context.Context, req *rpcpb.MakeContractTxReq,
+) (resp *rpcpb.MakeContractTxResp, err error) {
+	defer func() {
+		bytes, _ := json.Marshal(req)
+		if resp.Code != 0 {
+			logger.Warnf("make unsigned tx: %s error: %s", string(bytes), resp.RawMsgs)
```
Comment (Jerick26) [inline rpc/server/rpc_transaction.go]: resp.Message

**Pull request merged**